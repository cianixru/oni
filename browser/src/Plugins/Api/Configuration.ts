/**
 * Configuration.ts
 *
 * Implementation of configuration object
 *
 * Used for reading configuration values and listening to configuration changes
 */

import { Event, IEvent } from "./../../Event"

import * as Config from "./../../Config"

/**
 * API instance for interacting with Oni (and vim)
 */
export class Configuration {

    public get onConfigurationChangedEvent(): IEvent<void> {
        return this._onConfigurationChangedEvent
    }

    private _onConfigurationChangedEvent: Event<void> = new Event<void>("oni_configuration_changed")

    constructor() {
        Config.instance().onConfigChanged.subscribe(() => {
            this._onConfigurationChangedEvent.dispatch(null)
        })
    }

    public getValue<T>(configValue: string, defaultValue?: T): T {
        return Config.instance().getValue(configValue as any) || defaultValue
    }

}
