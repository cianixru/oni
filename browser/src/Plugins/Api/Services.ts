import { getInstance, VersionControlManager } from "./../../Services/VersionControl"

export class Services {
    public get vcs(): VersionControlManager {
        return getInstance()
    }
}
