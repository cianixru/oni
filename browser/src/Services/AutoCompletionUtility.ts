/**
 * AutoCompletionUtility.ts
 *
 * Helper functions for auto completion
 */

export function getCompletionStart(bufferLine: string, cursorColumn: number, completion: string): number {

    cursorColumn = Math.min(cursorColumn, bufferLine.length)

    let x = cursorColumn
    while (x >= 0) {

        const subWord = bufferLine.substring(x, cursorColumn + 1)

        if (completion.indexOf(subWord) === -1) {
            break
        }

        x--
    }

    return x + 1
}

export function replacePrefixWithCompletion(bufferLine: string, basePosition: number, cursorColumn: number, completion: string): string {
    const startPosition = basePosition

    const before = bufferLine.substring(0, startPosition)
    const after = bufferLine.substring(cursorColumn, bufferLine.length)

    return before + completion + after
}

export interface CompletionMeetResult {
    position: number
    base: string
    shouldExpandCompletions: boolean
}

export const doesCharacterMatchTriggerCharacters = (character: string, triggerCharacters: string[]): boolean => {
    return triggerCharacters.indexOf(character) >= 0
}

/**
 * Returns the start of the 'completion meet' along with the current base for completion
 */
export function getCompletionMeet(line: string, cursorColumn: number, characterMatchRegex: RegExp, completionTriggerCharacters: string[]): CompletionMeetResult {

    // Clamp column to within string bands
    let col = Math.max(cursorColumn - 1, 0)
    col = Math.min(col, line.length - 1)

    let currentPrefix = ""

    while (col >= 0 && col < line.length) {
        const currentCharacter = line[col]

        if (!currentCharacter.match(characterMatchRegex) || doesCharacterMatchTriggerCharacters(currentCharacter, completionTriggerCharacters)) {
            break
        }

        currentPrefix = currentCharacter + currentPrefix
        col--
    }

    const basePos = col

    const isFromTriggerCharacter = doesCharacterMatchTriggerCharacters(line[basePos], completionTriggerCharacters)

    const shouldExpandCompletions = currentPrefix.length > 0 || isFromTriggerCharacter

    // If the expansion is due to a letter, start the match at the letter position
    const position = isFromTriggerCharacter ? basePos : basePos + 1

    return {
        position: position + 1,
        base: currentPrefix,
        shouldExpandCompletions,
    }
}
