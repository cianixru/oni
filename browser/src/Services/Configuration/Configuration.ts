/**
 * Configuration.ts
 */

import * as Oni from "oni-api"
import { Event, IDisposable, IEvent } from "oni-types"
import { applyDefaultKeyBindings } from "./../../Input/KeyBindings"
import * as Log from "./../../Log"
import * as Performance from "./../../Performance"
import { diff } from "./../../Utility"

import { DefaultConfiguration } from "./DefaultConfiguration"
import { checkDeprecatedSettings } from "./DeprecatedConfigurationValues"
import { FileConfigurationProvider } from "./FileConfigurationProvider"
import { IConfigurationValues } from "./IConfigurationValues"
import { PersistedConfiguration } from "./PersistentSettings"
import * as UserConfiguration from "./UserConfiguration"

export interface IConfigurationProvider {
    onConfigurationChanged: IEvent<void>
    onConfigurationError: IEvent<Error>

    getValues(): GenericConfigurationValues
    getLastError(): Error | null

    activate(api: Oni.Plugin.Api): void
    deactivate(): void
}

export interface GenericConfigurationValues {
    [configKey: string]: any
}

interface ConfigurationProviderInfo {
    disposables: IDisposable[]
}

/**
 * Interface describing persistence layer for configuration
 */
export interface IPersistedConfiguration {
    getPersistedValues(): GenericConfigurationValues
    setPersistedValues(configurationValues: GenericConfigurationValues): void
}

export class Configuration implements Oni.Configuration {
    private _configurationProviders: IConfigurationProvider[] = []
    private _onConfigurationChangedEvent: Event<Partial<IConfigurationValues>> = new Event<
        Partial<IConfigurationValues>
    >()
    private _onConfigurationErrorEvent: Event<Error> = new Event<Error>()

    private _oniApi: Oni.Plugin.Api = null
    private _config: GenericConfigurationValues = {}

    private _setValues: { [configValue: string]: any } = {}
    private _fileToProvider: { [key: string]: IConfigurationProvider } = {}
    private _configProviderInfo = new Map<IConfigurationProvider, ConfigurationProviderInfo>()

    public get onConfigurationError(): IEvent<Error> {
        return this._onConfigurationErrorEvent
    }

    public get onConfigurationChanged(): IEvent<Partial<IConfigurationValues>> {
        return this._onConfigurationChangedEvent
    }

    constructor(
        private _defaultConfiguration: GenericConfigurationValues = DefaultConfiguration,
        private _persistedConfiguration: IPersistedConfiguration = new PersistedConfiguration(),
    ) {
        this._updateConfig()
    }

    public start(): void {
        Performance.mark("Config.load.start")

        this.addConfigurationFile(UserConfiguration.getUserConfigFilePath())

        Performance.mark("Config.load.end")
    }

    public addConfigurationFile(filePath: string): void {
        Log.info("[Configuration] Adding file: " + filePath)
        const fp = new FileConfigurationProvider(filePath)
        this.addConfigurationProvider(fp)
        this._fileToProvider[filePath] = fp
    }

    public removeConfigurationFile(filePath: string): void {
        Log.info("[Configuration] Removing file: " + filePath)
        const configProvider = this._fileToProvider[filePath]

        if (configProvider) {
            this.removeConfigurationProvider(configProvider)
            this._fileToProvider[filePath] = null
        }
    }

    public getErrors(): Error[] {
        return this._configurationProviders.map(cfp => cfp.getLastError())
    }

    public addConfigurationProvider(configurationProvider: IConfigurationProvider): void {
        this._configurationProviders.push(configurationProvider)

        const d1 = configurationProvider.onConfigurationChanged.subscribe(() => {
            Log.info("[Configuration] Got update.")
            this._updateConfig()
        })

        const d2 = configurationProvider.onConfigurationError.subscribe(error => {
            this._onConfigurationErrorEvent.dispatch(error)
        })

        this._configProviderInfo.set(configurationProvider, {
            disposables: [d1, d2],
        })

        this._updateConfig()
    }

    public removeConfigurationProvider(configurationProvider: IConfigurationProvider): void {
        this._configurationProviders = this._configurationProviders.filter(
            prov => prov !== configurationProvider,
        )

        const configurationInfo = this._configProviderInfo.get(configurationProvider)
        configurationInfo.disposables.forEach(dispose => dispose.dispose())

        this._configProviderInfo.delete(configurationProvider)

        this._updateConfig()
    }

    public hasValue(configValue: keyof IConfigurationValues): boolean {
        return !!this.getValue(configValue)
    }

    public setValues(configValues: { [configValue: string]: any }, persist: boolean = false): void {
        this._setValues = configValues

        this._config = {
            ...this._config,
            ...configValues,
        }

        if (persist) {
            this._persistedConfiguration.setPersistedValues(configValues)
        }

        this._onConfigurationChangedEvent.dispatch(configValues)
    }

    public getValue<K extends keyof IConfigurationValues>(configValue: K, defaultValue?: any) {
        if (typeof this._config[configValue] === "undefined") {
            return defaultValue
        } else {
            return this._config[configValue]
        }
    }

    public getValues(): GenericConfigurationValues {
        return { ...this._config }
    }

    public activate(oni: Oni.Plugin.Api): void {
        this._oniApi = oni

        this._activateIfOniObjectIsAvailable()
    }

    private _updateConfig(): void {
        const previousConfig = this._config

        let currentConfig = {
            ...this._defaultConfiguration,
            ...this._persistedConfiguration.getPersistedValues(),
            ...this._setValues,
        }

        this._configurationProviders.forEach(configProvider => {
            const configurationValues = configProvider.getValues()
            currentConfig = { ...currentConfig, ...configurationValues }
        })

        this._config = currentConfig

        checkDeprecatedSettings(this._config)

        this._deactivate()
        this._activateIfOniObjectIsAvailable()

        this._notifyListeners(previousConfig)
    }

    private _activateIfOniObjectIsAvailable(): void {
        if (this._oniApi) {
            applyDefaultKeyBindings(this._oniApi, this)
            this._configurationProviders.forEach(configurationProvider =>
                configurationProvider.activate(this._oniApi),
            )
        }
    }

    private _deactivate(): void {
        this._configurationProviders.forEach(configurationProvider =>
            configurationProvider.deactivate(),
        )
        if (this._config && this._config.deactivate) {
            this._config.deactivate()
        }
    }

    private _notifyListeners(previousConfig?: Partial<IConfigurationValues>): void {
        previousConfig = previousConfig || {}

        const changedValues = diff(this._config, previousConfig)

        const diffObject = changedValues.reduce(
            (previous: Partial<IConfigurationValues>, current: string) => {
                const currentValue = this._config[current]

                // Skip functions, because those will always be different
                if (currentValue && typeof currentValue === "function") {
                    return previous
                }

                return {
                    ...previous,
                    [current]: this._config[current],
                }
            },
            {},
        )

        this._onConfigurationChangedEvent.dispatch(diffObject)
    }
}

export const configuration = new Configuration()
