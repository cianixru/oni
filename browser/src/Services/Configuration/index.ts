export * from "./Configuration"
export * from "./IConfigurationValues"
export * from "./UserConfiguration"
