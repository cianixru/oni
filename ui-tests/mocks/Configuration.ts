import * as Oni from "oni-api"

const Configuration = jest.fn<Oni.Configuration>().mockImplementation(() => ({
    onConfigurationChanged() {
        return {
            subscribe: jest.fn(),
        }
    },
    notifyListeners: jest.fn(),
    updateConfig: jest.fn(),
    getValue: jest.fn(),
}))

export const configuration = new Configuration()

export default Configuration
