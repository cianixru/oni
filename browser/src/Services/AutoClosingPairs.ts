/**
 * AutoClosingPairs
 *
 * Service to enable auto-closing pair key bindings
 */

import * as Oni from "oni-api"

import { Configuration } from "./Configuration"
import { EditorManager } from "./EditorManager"
import { InputManager } from "./InputManager"
import { LanguageManager } from "./Language"

import { NeovimInstance } from "./../neovim"

import * as Log from "./../Log"

export interface IAutoClosingPair {
    open: string
    close: string
    // TODO: Support `notIn` equivalent
}

export const activate = (
    configuration: Configuration,
    editorManager: EditorManager,
    inputManager: InputManager,
    languageManager: LanguageManager,
) => {
    const insertModeFilter = () => editorManager.activeEditor.mode === "insert"

    let subscriptions: Oni.DisposeFunction[] = []

    const handleOpenCharacter = (pair: IAutoClosingPair, editor: Oni.Editor) => () => {
        const neovim: NeovimInstance = editor.neovim as any
        neovim.blockInput(async (inputFunc: any) => {
            // TODO: PERFORMANCE: Look at how to collapse this instead of needed multiple asynchronous calls.
            await inputFunc(pair.open + pair.close)

            const pos = await neovim.callFunction("getpos", ["."])
            const [, oneBasedLine, oneBasedColumn] = pos
            await editor.activeBuffer.setCursorPosition(oneBasedLine - 1, oneBasedColumn - 2)
        })

        return true
    }

    const handleBackspaceCharacter = (pairs: IAutoClosingPair[], editor: Oni.Editor) => () => {
        const neovim: NeovimInstance = editor.neovim as any
        neovim.blockInput(async (inputFunc: any) => {
            const activeBuffer = editor.activeBuffer
            const lines = await activeBuffer.getLines(
                activeBuffer.cursor.line,
                activeBuffer.cursor.line + 1,
            )
            const line = lines[0]

            const { column } = activeBuffer.cursor

            const matchingPair = pairs.find(p => {
                return column >= 1 && line[column] === p.close && line[column - 1] === p.open
            })

            if (matchingPair) {
                // Remove the pairs
                const beforePair = line.substring(0, column - 1)
                const afterPair = line.substring(column + 1, line.length)

                const pos = await neovim.callFunction("getpos", ["."])
                const [, oneBasedLine, oneBasedColumn] = pos
                await editor.activeBuffer.setCursorPosition(oneBasedLine - 1, oneBasedColumn - 2)

                await activeBuffer.setLines(
                    activeBuffer.cursor.line,
                    activeBuffer.cursor.line + 1,
                    [beforePair + afterPair],
                )
            } else {
                await inputFunc("<bs>")
            }
        })

        return true
    }

    const handleEnterCharacter = (pairs: IAutoClosingPair[], editor: Oni.Editor) => () => {
        const neovim: NeovimInstance = editor.neovim as any
        neovim.blockInput(async (inputFunc: any) => {
            const activeBuffer = editor.activeBuffer

            const lines = await (activeBuffer as any).getLines(
                activeBuffer.cursor.line,
                activeBuffer.cursor.line + 1,
                false,
            )
            const line = lines[0]

            const { column } = activeBuffer.cursor

            const matchingPair = pairs.find(p => {
                return column >= 1 && line[column] === p.close && line[column - 1] === p.open
            })

            if (matchingPair) {
                const whiteSpacePrefix = getWhiteSpacePrefix(line)
                const beforePair = line.substring(0, column)
                const afterPair = line.substring(column, line.length)

                const pos = await neovim.callFunction("getpos", ["."])
                const [, oneBasedLine] = pos
                await activeBuffer.setLines(
                    activeBuffer.cursor.line,
                    activeBuffer.cursor.line + 1,
                    [beforePair, whiteSpacePrefix, whiteSpacePrefix + afterPair],
                )
                await activeBuffer.setCursorPosition(oneBasedLine, whiteSpacePrefix.length)
                await inputFunc("<tab>")
            } else {
                await inputFunc("<enter>")
            }
        })

        return true
    }

    const handleCloseCharacter = (pair: IAutoClosingPair, editor: Oni.Editor) => () => {
        const neovim: any = editor.neovim
        neovim.blockInput(async (inputFunc: any) => {
            const activeBuffer = editor.activeBuffer
            const lines = await (activeBuffer as any).getLines(
                activeBuffer.cursor.line,
                activeBuffer.cursor.line + 1,
                false,
            )
            const line = lines[0]
            if (line[activeBuffer.cursor.column] === pair.close) {
                await activeBuffer.setCursorPosition(
                    activeBuffer.cursor.line,
                    activeBuffer.cursor.column + 1,
                )
            } else {
                await inputFunc(pair.close)
            }
        })

        return true
    }

    const onBufferEnter = (newBuffer: Oni.Buffer) => {
        if (!configuration.getValue("autoClosingPairs.enabled")) {
            Log.verbose("[Auto Closing Pairs] Not enabled.")
            return
        }

        if (subscriptions && subscriptions.length) {
            subscriptions.forEach(df => df())
        }

        subscriptions = []

        const autoClosingPairs = getAutoClosingPairs(configuration, newBuffer.language)

        autoClosingPairs.forEach(pair => {
            subscriptions.push(
                inputManager.bind(
                    pair.open,
                    handleOpenCharacter(pair, editorManager.activeEditor),
                    insertModeFilter,
                ),
            )
            subscriptions.push(
                inputManager.bind(
                    pair.close,
                    handleCloseCharacter(pair, editorManager.activeEditor),
                    insertModeFilter,
                ),
            )
        })

        subscriptions.push(
            inputManager.bind(
                "<bs>",
                handleBackspaceCharacter(autoClosingPairs, editorManager.activeEditor),
                insertModeFilter,
            ),
        )
        subscriptions.push(
            inputManager.bind(
                "<enter>",
                handleEnterCharacter(autoClosingPairs, editorManager.activeEditor),
                insertModeFilter,
            ),
        )
    }

    editorManager.activeEditor.onBufferEnter.subscribe(onBufferEnter)

    const activeEditor = editorManager.activeEditor
    if (activeEditor && activeEditor.activeBuffer) {
        onBufferEnter(activeEditor.activeBuffer)
    }
}

const nonWhiteSpaceRegEx = /\S/

export const getWhiteSpacePrefix = (str: string): string => {
    const firstIndex = str.search(nonWhiteSpaceRegEx)

    if (firstIndex === -1) {
        return ""
    } else {
        return str.substring(0, firstIndex)
    }
}

const getAutoClosingPairs = (
    configuration: Configuration,
    language: string,
): IAutoClosingPair[] => {
    const languagePairs = configuration.getValue(`language.${language}.autoClosingPairs`)

    if (languagePairs) {
        return languagePairs
    } else {
        return configuration.getValue("autoClosingPairs.default") || []
    }
}
