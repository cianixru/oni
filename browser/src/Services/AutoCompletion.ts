/**
 * AutoCompletion.ts
 *
 * Manages the autocompletion service:
 *  - Integrating with Neovim's external popup menu
 *  - Managing completion state
 */

import * as types from "vscode-languageserver-types"

import { IBuffer, INeovimInstance } from "./../neovim"

import * as Utility from "./AutoCompletionUtility"

import * as UI from "./../UI/index"

export class AutoCompletion {

    constructor(
        private _neovimInstance: INeovimInstance,
    ) {
        this._neovimInstance.on("show-popup-menu", (completions: any[]) => {
            const c = completions.map((completion) => ({
                kind: types.CompletionItemKind.Text,
                label: completion[0],
            }))

            UI.Actions.showCompletions({
                base: "",
                completions: c,
            })
        })

        this._neovimInstance.on("hide-popup-menu", () => {
            UI.Actions.hideCompletions()
        })
    }

    public complete(): void {
        let cursorRow: number
        let cursorColumn: number
        let originalLineLength: number
        let newLineLength: number

        const completion = UI.Selectors.getSelectedCompletion() || ""
        let currentBuffer: IBuffer
        this._neovimInstance.getCurrentBuffer()
            .then((buffer) => currentBuffer = buffer)

            // Get position in buffer
            .then(() => this._neovimInstance.getCursorColumn())
            .then((col) => cursorColumn = col)

            .then(() => this._neovimInstance.getCursorRow())
            .then((row) => cursorRow = row)

            // Get current line
            .then(() => {
                return currentBuffer.getLines(cursorRow - 1, cursorRow, false)
            })
            // Replace with completion value
            .then((value) => {
                const line = value[0]
                originalLineLength = line.length
                const newLine = Utility.replacePrefixWithCompletion(line, cursorColumn - 2, completion)
                newLineLength = newLine.length
                return currentBuffer.setLines(cursorRow - 1, cursorRow, false, [newLine])
            })
            .then(() => {
                const cursorOffset = newLineLength - originalLineLength

                // Set cursor position after the word
                return this._neovimInstance.eval(`setpos(".", [0, ${cursorRow}, ${cursorColumn + cursorOffset}, 0])`)
            })

        UI.Actions.hideCompletions()
    }
}
