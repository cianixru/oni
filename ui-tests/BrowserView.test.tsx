import { shallow } from "enzyme"
import { shallowToJson } from "enzyme-to-json"
import * as React from "react"

import { Event } from "oni-types"
import {
    BrowserView,
    IBrowserViewProps,
    IBrowserViewState,
} from "../browser/src/Services/Browser/BrowserView"
import { Configuration } from "../browser/src/Services/Configuration"

const mockEvent = new Event<void>()

// Using the disable life cycle methods here as the CDM calls sneak which is
// an external dependency I'm not trying to test here

describe("<BrowserView /> Tests", () => {
    const component = (
        <BrowserView
            initialUrl={"test.com"}
            configuration={{} as Configuration}
            debug={mockEvent}
            goBack={mockEvent}
            goForward={mockEvent}
            reload={mockEvent}
        />
    )
    it("component to render correctly", () => {
        const wrapper = shallow(component, { disableLifecycleMethods: true })
        expect(wrapper).toBeDefined()
    })
    it('Should correctly prefix a url with "https" or "http" if needed', () => {
        const wrapper = shallow(component, { disableLifecycleMethods: true })
        // can be typed here but the _webviewComponent is expected - TS Error
        const instance: any = wrapper.instance()
        expect(instance.prefixUrl("apple.com")).toBe("http://apple.com")
    })
    it('Should NOT prefix a url with "https" or "http" if already present', () => {
        const wrapper = shallow(component, { disableLifecycleMethods: true })
        const instance: any = wrapper.instance()
        // subtle difference here as the function always add https as a prefix not http
        expect(instance.prefixUrl("https://www.apple.com")).toBe("https://www.apple.com")
    })

    it("Should match the recent snapshot - unless an intentional change has occurred ", () => {
        const wrapper = shallow(component, { disableLifecycleMethods: true })
        expect(shallowToJson(wrapper)).toMatchSnapshot()
    })
})
