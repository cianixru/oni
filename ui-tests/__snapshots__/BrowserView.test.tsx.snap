// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrowserView /> Tests Should match the recent snapshot - unless an intentional change has occurred  1`] = `
<styled.div
  key="test2"
>
  <styled.div>
    <Component
      icon="chevron-left"
      onClick={[Function]}
    />
    <Component
      icon="chevron-right"
      onClick={[Function]}
    />
    <Component
      icon="undo"
      onClick={[Function]}
    />
    <AddressBarView
      onAddressChanged={[Function]}
      url="test.com"
    />
    <Component
      icon="bug"
      onClick={[Function]}
    />
  </styled.div>
  <styled.div>
    <div
      key="test"
      style={
        Object {
          "bottom": "0px",
          "left": "0px",
          "position": "absolute",
          "right": "0px",
          "top": "0px",
        }
      }
    />
  </styled.div>
</styled.div>
`;
